### PR TITLE
Replace wstring used in public methods by Platform::String in CalcViewModel

### LINKED ISSUE #757: Prefer Platform::String over wstring in CalcViewModel
**Describe the bug**
CalcViewModel currently uses wstring and Platform::String to represent strings. We should prefer Platform::String to make it easier to interact with XAML and WinRT api and reduce the number of conversions between the 2 formats in the Calculator project.

**Requested Assignment**
If possible, I would like to fix this.

### PR DESCRIPTION
**Pull request by rudyhuyn**
### Description of the changes:
- replace all wstring from public methods to `Platform::String`

also
- replace std::task by `IAsyncOperation`
- convert `LocalizationStringUtil` and `AppResourceProvider` to "ref classes" 
- use wstringstream when we concatenate many strings.

### How changes were validated:
- manually

### CODE REVIEW

File: src/CalcViewModel/Common/LocalizationStringUtil.h [hunk 1]
```diff
@@ -9,50 +9,87 @@ namespace CalculatorApp
 {
     namespace Common
     {
-        class LocalizationStringUtil
+         class LocalizationStringUtilInternal
```
Comment (mcooley) [inline src/CalcViewModel/Common/LocalizationStringUtil.h]: nit: extra space

File: src/CalcViewModel/Common/LocalizationStringUtil.h [hunk 2]
```diff
@@ -9,50 +9,87 @@ namespace CalculatorApp
...
         {
         public:
-            static std::wstring GetLocalizedString(const wchar_t* pMessage, ...)
-            {
+            static Platform::String ^ GetLocalizedString(Platform::String ^ pMessage, ...) {
```
Comment (mcooley) [inline src/CalcViewModel/Common/LocalizationStringUtil.h]: nit: brace on new line

File: src/CalcViewModel/Common/LocalizationStringUtil.h [hunk 3]
```diff
@@ -9,50 +9,87 @@ namespace CalculatorApp
...
+                ^ GetLocalizedString(Platform::String ^ pMessage)
+                {
+                    return LocalizationStringUtilInternal::GetLocalizedString(pMessage);
+                }
 
-        private:
-            static void EnsureInitialization(Platform::String^ resourceKey, Platform::String^& formatVariable)
-            {
-                if (resourceKey == nullptr || resourceKey->IsEmpty())
+                static Platform::String
```
Comment (mcooley) [inline src/CalcViewModel/Common/LocalizationStringUtil.h]: nit: all methods in this class have weird formatting

File: src/CalcViewModel/DataLoaders/CurrencyDataLoader.cpp [hunk 4]
```diff
@@ -753,15 +755,15 @@ wstring CurrencyDataLoader::GetCurrencyTimestamp()
...
         DateTimeFormatter ^ timeFormatter = ref new DateTimeFormatter(L"shorttime");
-        wstring time = timeFormatter->Format(m_cacheTimestamp)->Data();
+        auto time = timeFormatter->Format(m_cacheTimestamp);
 
-        timestamp = LocalizationStringUtil::GetLocalizedString(m_timestampFormat.c_str(), date.c_str(), time.c_str());
+        return LocalizationStringUtil::GetLocalizedString(m_timestampFormat, date, time)->Data();
     }
 
-    return timestamp;
+    return L"";
```
Comment (mcooley) [inline src/CalcViewModel/DataLoaders/CurrencyDataLoader.cpp]: Local variable timestamp is unused and can be removed

File: src/CalcViewModel/HistoryViewModel.h [hunk 5]
```diff
@@ -68,7 +68,7 @@ namespace CalculatorApp
             void UpdateHistoryVectorLength(_In_ int newValue, _In_ CalculationManager::CALCULATOR_MODE cMode);
             bool IsValid(_In_ CalculationManager::HISTORYITEM item);
 
-            void MakeHistoryClearedNarratorAnnouncement(Platform::String ^ resourceKey, Platform::String ^ &formatVariable);
+            Platform::String ^ MakeHistoryClearedNarratorAnnouncement(Platform::String ^ resourceKey, Platform::String ^ formatVariable);
```
Comment (mcooley) [inline src/CalcViewModel/HistoryViewModel.h]: I don't think this change is necessary?

File: src/CalcViewModel/UnitConverterViewModel.cpp [hunk 6]
```diff
@@ -515,8 +515,10 @@ void UnitConverterViewModel::OnPasteCommand(Platform::Object ^ parameter)
     // Ensure that the paste happens on the UI thread
     // EventWriteClipboardPaste_Start();
     // Any converter ViewMode is fine here.
-    CopyPasteManager::GetStringToPaste(m_Mode, NavCategory::GetGroupType(m_Mode))
-        .then([this](String ^ pastedString) { OnPaste(pastedString); }, concurrency::task_continuation_context::use_current());
+    
+    auto that(this);
```
Comment (mcooley) [inline src/CalcViewModel/UnitConverterViewModel.cpp]: Instead can you lambda capture this by value? c++17 feature, I'm not sure if it works with CX.

[*this](String ^ pastedString) { this->... }

Comment (mcooley) [reply src/CalcViewModel/UnitConverterViewModel.cpp]: Same feedback for UnitConverter.xaml.cpp:246

File: src/CalcViewModel/Common/CopyPasteManager.cpp [hunk 7]
```diff
@@ -76,36 +76,30 @@ task<String ^> CopyPasteManager::GetStringToPaste(ViewMode mode, CategoryGroupTy
...
 String ^ CopyPasteManager::ValidatePasteExpression(String ^ pastedText, ViewMode mode, int programmerNumberBase, BitLength bitLengthType)
 {
-    return CopyPasteManager::ValidatePasteExpression(pastedText, mode, NavCategory::GetGroupType(mode), programmerNumberBase, bitLengthType);
+    return ValidatePasteExpression(pastedText, mode, NavCategory::GetGroupType(mode), programmerNumberBase, bitLengthType);
 }
 
 // return "NoOp" if pastedText is invalid else return pastedText
-
-String ^ CopyPasteManager::ValidatePasteExpression(
+String
```
[the remark this answers is not in the document]
Comment (rudyhuyn) [reply src/CalcViewModel/Common/CopyPasteManager.cpp]: 559ED56C-B63E-4F8B-B85C-299B24C2BB0F


Our CLang format rules are not great with C++/CX, we have a lot of formatting problems with it (when using `^`, `internal`, etc...). We should take the time to study how we can solve this problem.

File: src/CalcViewModel/Common/LocalizationStringUtil.h (hunk 3 above)
Comment (rudyhuyn) [reply src/CalcViewModel/Common/LocalizationStringUtil.h]: A good example of how broken is clang with C++/CX  (especially with `^`) 😫

File: src/CalcViewModel/HistoryViewModel.h (hunk 5 above)
Comment (rudyhuyn) [reply src/CalcViewModel/HistoryViewModel.h]: Unfortunately, this change is necessary for the long-term.

File: src/CalcViewModel/UnitConverterViewModel.cpp (hunk 6 above)
Comment (rudyhuyn) [reply src/CalcViewModel/UnitConverterViewModel.cpp]: I tried but it's not possible :(
`error C3076: you cannot embed an instance of a reference type`, same with `[&]` and `[=]`.

`that` is the only workaround I know at the moment.

**Pull request merged**